### PR TITLE
[python] Workaround MSVC2022-related bug

### LINKED ISSUE #2090: python_wrap.cxx build failure with VS2022 Preview
When building with VS2022 Preview 4.0 onwards, I get a bunch of errors when compiling python wrapper class which was generated with SWIG 4.0.2 on Windows. This compiles fine with VS2019. Using Python 3.9.7 64bit.

```
[777/779] Building CXX object ui\python\CMakeFiles\py3.dir\3\x86_64\PYTHON_wrap.cxx.obj
  FAILED: ui/python/CMakeFiles/py3.dir/3/x86_64/PYTHON_wrap.cxx.obj 
  C:\PROGRA~1\MIB055~1\2022\Preview\VC\Tools\MSVC\1430~1.307\bin\Hostx64\x64\cl.exe  /nologo /TP -DSWIG_PYTHON_INTERPRETER_NO_DEBUG -D_CRT_SECURE_NO_WARNINGS -Dpy3_EXPORTS -IC:\python\include ...includes removed... /DWIN32 /D_WINDOWS /EHsc /sdl /GR /EHsc /DNOMINMAX /Zi /Ob0 /Od /RTC1 /Od -MDd /W4 /WX /MP /bigobj /WX- /wd5033 -std:c++latest /showIncludes /Foui\python\CMakeFiles\py3.dir\3\x86_64\PYTHON_wrap.cxx.obj /Fdui\python\CMakeFiles\py3.dir\ /FS -c D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx
D:\src\out\build\x64-Debug\cl : Command line warning D9025: overriding '/WX' with '/WX-'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xmemory(154): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xmemory(154): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xmemory(164): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xmemory(164): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xmemory(1203): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xmemory(1203): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xpolymorphic_allocator.h(146): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xpolymorphic_allocator.h(146): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xpolymorphic_allocator.h(152): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\xpolymorphic_allocator.h(152): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(206): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(206): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(221): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(221): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(237): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(237): error C3861: '_invalid_parameter': identifier not found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(297): error C2039: '_invalid_parameter': is not a member of '`global namespace''
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.30.30704\include\atomic(297): error C3861: '_invalid_parameter': identifier not found
D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4492): warning C4127: conditional expression is constant
  D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4492): note: consider using 'if constexpr' statement instead
D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4498): warning C4127: conditional expression is constant
  D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4498): note: consider using 'if constexpr' statement instead
D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4525): warning C4127: conditional expression is constant
  D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4525): note: consider using 'if constexpr' statement instead
D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4634): warning C4127: conditional expression is constant
  D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4634): note: consider using 'if constexpr' statement instead
D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4640): warning C4127: conditional expression is constant
  D:\src\ui\python\3\x86_64\PYTHON_wrap.cxx(4640): note: consider using 'if constexpr' statement instead
```

### PR DESCRIPTION
**Pull request by ojwb**
Fixes #2090

### CODE REVIEW

**Comment by wsfulton** [comment]
I wonder if users should not just use %begin to add this header in. What is the Python dev community take on this? If Visual Studio 2022 is not supported by Python, I don't think SWIG should support it either for Python wrappers.

**Comment by ojwb** [comment]
> I wonder if users should not just use %begin to add this header in.

That's the other option I gave in the issue.  Mostly I felt we should either add the workaround or close the issue with a clear statement that people should workaround - not working around and keeping the issue open helps nobody.

There's already "prior art" for us working around such things though (an issue with `hypot()`), and it seems better to fix this in one place in SWIG rather than forcing everyone to rediscover the fix for themselves.

It does mean carrying a small extra conditional indefinitely (at least until someone determines it isn't needed any more).

> What is the Python dev community take on this? 

I'm not really a Python dev so I don't really know.  I couldn't find a bug report of this issue.

> If Visual Studio 2022 is not supported by Python, I don't think SWIG should support it either for Python wrappers.

According to

https://wiki.python.org/moin/WindowsCompilers#Which_Microsoft_Visual_C.2B-.2B-_compiler_to_use_with_a_specific_Python_version_.3F

for Python 3.5, 3.6, 3.7, 3.8 (nothing newer listed) you should use 14.x and 2022 seems to be 14.30 from the quoted output in the issue.  Not 100% clear cut, but I'd argue that can be reasonably interpreted as "supported".

**Comment by vadz** [comment]
FWIW I definitely think SWIG should provide a workaround out of the box rather than forcing everybody to do it in their code. What possible reason could there be _not_ to do this?

**Comment by ojwb** [comment]
This seems similar to #325.  The suggested workaround there seems to be `#include <crtdefs.h>` and it's only needed in non-debug mode, but I wonder if it's actually the same problem underneath.  But it's not a platform or scripting language I actively use.

**Comment by ojwb** [comment]
Merged, since opinion seemed to lean that way.

File: Lib/python/pyruntime.swg [hunk 1]
```diff
@@ -4,6 +4,13 @@
 # include <math.h>
 #endif
 
+#if defined(_MSC_VER) && _MSC_VER >= 1930
```
Comment (montant) [inline Lib/python/pyruntime.swg:7]: bug #2174 can be observed already with vs 2019 16.11 (not with 16.11.5, but it is present with 16.11.17 - can't say with which update it appeared in between). As consequence, if we want to activate fix for vs 2019 16.11, test on _MSC_VER should rather be >= 1929

Comment (ojwb) [reply Lib/python/pyruntime.swg:7]: @montant Please open a PR with your proposed change so it gets tested automatically by CI.

**Comment by wsfulton** [comment]
@montant you must be using defining SWIG_PYTHON_INTERPRETER_NO_DEBUG as mentioned in [docs](https://swig.org/Doc4.0/Python.html#Python_nn12) ? More discussion on this and other improvements in #2090.

**Comment by montant** [comment]
> @montant you must be using defining SWIG_PYTHON_INTERPRETER_NO_DEBUG as mentioned in [docs](https://swig.org/Doc4.0/Python.html#Python_nn12) ? More discussion on this and other improvements in #2090.

I confirm: SWIG_PYTHON_INTERPRETER_NO_DEBUG is defined in my case.

**Pull request merged**